### DISCUSSION TITLE
Cache stampede problem in `CacheBase::getOrSet()`

### DISCUSSION THREAD

**Discussion by ardenwick** [ClickHouse/ClickHouse#54648] [Q&A] [answered]
In `CacheBase::getOrSet()` from [CacheBase.h](https://github.com/ClickHouse/ClickHouse/blob/master/src/Common/CacheBase.h#L119):

https://github.com/ClickHouse/ClickHouse/blob/de9662f7a4916a46d25dbb03b16d0220603ae1fb/src/Common/CacheBase.h#L110-L119

It addresses **cache stampede** problem with 'InsertToken' mechanism, but I'm confused by code logic in the function, it looks complicated, or is there a corresponding design pattern behind it ?

Here if I simply wrap all code logic inside a locked mutex, could it be 'cache stampede preventing' ?

```C++
    template <typename LoadFunc>
    std::pair<MappedPtr, bool> getOrSet(const Key & key, LoadFunc && load_func)
    {
        std::lock_guard lock(mutex); // no concurrent calls to load_func() will occur, because of lock protection

        auto val = cache_policy->get(key, lock);
        if (val)
        {
            ++hits;
            return std::make_pair(val, false);
        }

        val = load_func();
        cache_policy->set(key, val, lock);

        ++misses;
        return std::make_pair(val, true);
    }

```

Or should I think the other way: the `cache_policy->set()` **can be heavy**, so the 'InsertToken' mechanism is made such to avoid waiting for `cache_policy->set()` to return, and thus improve performance of concurrent 'getOrSet()' calls ?

**Comment by alexey-milovidov** [comment, marked as answer]
Concurrent threads getting or refreshing data for the same key should wait for each other, so only one thread will refresh the data (which can be costly), and all the other threads will retrieve the data from the cache.

But concurrent threads getting or refreshing data for the different keys should not block each other - they should do these operations concurrently. Yes, calculating the data that is not in the cache can be costly.

**Reply by ardenwick** [reply]
Thanks for the answer!
